File: ios/RNTrackPlayer/Logic/MediaWrapper.m
//
//  MediaWrapper.m
//  RNTrackPlayer
//
//  Created by David Chavez on 7/2/17.
//  Copyright © 2017 David Chavez. All rights reserved.
//

@import MediaPlayer;

#import "MediaWrapper.h"
#import "STKAudioPlayer.h"

@interface MediaWrapper() <STKAudioPlayerDelegate>
@property(strong, nonatomic) NSMutableArray *queue;
@property(strong, nonatomic) STKAudioPlayer *player;
@property(assign, nonatomic) NSInteger currentIndex;
@property(strong, nonatomic) NSURLSessionDataTask *imageDownloader;
@end

@implementation MediaWrapper

- (id)init {
    self = [super init];
    if (self) {
        _currentIndex = 0;
        _queue = [[NSMutableArray alloc] init];
        _player = [[STKAudioPlayer alloc] init];
        
        [[UIApplication sharedApplication] beginReceivingRemoteControlEvents];
    }
    
    return self;
}

- (void)dealloc {
    [_player dispose];
}


// MARK: - Public Interface

- (BOOL)queueContainsTrack:(NSString *)trackId {
    for (id track in _queue) {
        if ([[(Track *)track identifier] isEqualToString:trackId]) {
            return YES;
        }
    }
    
    return NO;
}

- (void)addTracks:(NSArray *)tracks before:(NSString *)trackId {
    __block BOOL indexFound = NO;
    __block NSInteger indexToInsertAt = 0;
    
    [_queue enumerateObjectsUsingBlock:^(id track, NSUInteger index, BOOL *stop) {
        if ([[(Track *)track identifier] isEqualToString:trackId]) {
            indexFound = YES;
            indexToInsertAt = indexToInsertAt;
        }
    }];
    
    if (indexFound) {
        NSRange range = NSMakeRange(indexToInsertAt, tracks.count);
        NSIndexSet *indexes = [NSIndexSet indexSetWithIndexesInRange:range];
        [_queue insertObjects:tracks atIndexes:indexes];
    } else {
        [_queue addObjectsFromArray:tracks];
    }
}

- (void)removeTrackIds:(NSArray *)trackIds {
    __block BOOL removedCurrentTrack = NO;
    
    NSUInteger numberOfTracksToKeep = _queue.count - trackIds.count;
    __block NSMutableArray *tracksToKeep = [NSMutableArray arrayWithCapacity:numberOfTracksToKeep];
    
    [_queue enumerateObjectsUsingBlock:^(id track, NSUInteger queueIndex, BOOL *stop) {
        for (id identifier in trackIds) {
            if ([[(Track *)track identifier] isEqualToString:(NSString *)identifier]) {
                if (queueIndex == _currentIndex) { removedCurrentTrack = YES; }
                else if (queueIndex < _currentIndex) { _currentIndex--; }
            } else {
                [tracksToKeep addObject:track];
            }
        }
    }];
    
    if (removedCurrentTrack) {
        if (_currentIndex > _queue.count - 1) {
            _currentIndex = _queue.count;
            [self stop];
        } else {
            [self play];
        }
    }
    
    [_queue setArray:tracksToKeep];
}

- (void)skipToTrack:(NSString *)trackId {
    [_queue enumerateObjectsUsingBlock:^(id track, NSUInteger index, BOOL *stop) {
        if ([[(Track *)track identifier] isEqualToString:(NSString *)trackId]) {
            _currentIndex = index;
            [self play];
        }
    }];
}

- (BOOL)playNext {
    if (_currentIndex < _queue.count - 1) {
        _currentIndex++;
        [self play];
        return YES;
    } else {
        [self stop];
        return NO;
    }
}

- (BOOL)playPrevious {
    if (_currentIndex > 0) {
        _currentIndex--;
        [self play];
        return YES;
    }
    
    return NO;
}

- (void)reset {
    [_queue removeAllObjects];
    [_player stop];
}

- (void)play {
    // Resume playback if it was paused
    if (_player.state == STKAudioPlayerStatePaused) {
        [_player resume];
        return;
    }
    
    Track *track = [_queue objectAtIndex:_currentIndex];
    [_imageDownloader cancel];
    
    // TODO: Handle artwork (part of supporting local files)
    NSDictionary *metadata = @{
                               MPMediaItemPropertyMediaType: @(MPMediaTypeMusic),
                               MPMediaItemPropertyTitle: track.title ? track.title : @"",
                               MPMediaItemPropertyAlbumTitle: track.album ? track.album : @"",
                               MPMediaItemPropertyArtist: track.artist ? track.artist : @"",
                               MPMediaItemPropertyGenre: track.genre ? track.genre : @"",
                               MPMediaItemPropertyPlaybackDuration: track.duration ? track.duration : @(0),
                               MPMediaItemPropertyReleaseDate: track.date ? track.date : @"",
                               MPNowPlayingInfoPropertyElapsedPlaybackTime: @(_player.progress),
                               };
    
    [[MPNowPlayingInfoCenter defaultCenter] setNowPlayingInfo:metadata];
    
    _imageDownloader = [[NSURLSession sharedSession]
                        dataTaskWithURL:track.artwork.value
                        completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                            if (!error) {
                                NSMutableDictionary *dict = [MPNowPlayingInfoCenter defaultCenter].nowPlayingInfo.mutableCopy;
                                UIImage *artwork = [UIImage imageWithData:data];
                                dict[MPMediaItemPropertyArtwork] = [[MPMediaItemArtwork alloc] initWithImage:artwork];
                                [[MPNowPlayingInfoCenter defaultCenter] setNowPlayingInfo:dict];
                            }
                        }];
    [_imageDownloader resume];
    
    // fetch possible redirected url
    if (!track.url.isLocal) {
        NSMutableURLRequest *newRequest = [[NSMutableURLRequest alloc] initWithURL:track.url.value];
        NSURLResponse *response = nil;
        NSError *error = nil;
        
        [newRequest setValue:@"HEAD" forKey:@"HTTPMethod"];
        [NSURLConnection
         sendSynchronousRequest:newRequest
         returningResponse:&response
         error:&error];
        
        if (response && !error) {
            [track.url setValue:[response URL]];
        }
    }
    
    [_player playURL:track.url.value];
}

- (void)pause {
    [_player pause];
}

- (void)stop {
    [_player stop];
    [[MPNowPlayingInfoCenter defaultCenter] setNowPlayingInfo:nil];
}

- (void)seekToTime:(NSInteger)time {
    double delayInSeconds = 0.1;
    int64_t delay = (int64_t)(delayInSeconds * NSEC_PER_SEC);
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, delay);
    dispatch_after(popTime, dispatch_get_main_queue(), ^() {
        [_player seekToTime:time];
    });
}

- (void)setVolume:(NSInteger)volume {
    [_player setVolume:volume];
}

- (Track *)currentTrack {
    return _queue[_currentIndex];
}

- (NSInteger)duration {
    return _player.duration;
}

- (NSInteger)position {
    return _player.progress;
}

- (NSString *)state {
    // Mapping to values used by Android project
    // TODO: - Unify into a shared enum
    switch (_player.state) {
        case STKAudioPlayerStatePlaying:
            return @"STATE_PLAYING";
        case STKAudioPlayerStatePaused:
            return @"STATE_PAUSED";
        case STKAudioPlayerStateStopped:
            return @"STATE_STOPPED";
        case STKAudioPlayerStateBuffering:
            return @"STATE_BUFFERING";
        default:
            return @"STATE_NONE";
    }
}


// MARK: - STKAudioPlayerDelegate

- (void)audioPlayer:(STKAudioPlayer *)audioPlayer
didFinishBufferingSourceWithQueueItemId:(NSObject *)queueItemId {}

- (void)audioPlayer:(STKAudioPlayer *)audioPlayer
didFinishPlayingQueueItemId:(NSObject *)queueItemId
         withReason:(STKAudioPlayerStopReason)stopReason
        andProgress:(double)progress
        andDuration:(double)duration {
    // Play the next song in the queue when song ends
    if (stopReason == STKAudioPlayerStopReasonEof) {
        [self playNext];
        [_imageDownloader cancel];
        [[MPNowPlayingInfoCenter defaultCenter] setNowPlayingInfo:@{}];
    }
}

- (void)audioPlayer:(STKAudioPlayer *)audioPlayer
didStartPlayingQueueItemId:(NSObject *)queueItemId {}

- (void)audioPlayer:(STKAudioPlayer *)audioPlayer
       stateChanged:(STKAudioPlayerState)state
      previousState:(STKAudioPlayerState)previousState {}

- (void)audioPlayer:(STKAudioPlayer *)audioPlayer
    unexpectedError:(STKAudioPlayerErrorCode)errorCode {}

@end

File: ios/RNTrackPlayer/Logic/MediaWrapper.swift
//
//  MediaWrapper.swift
//  RNTrackPlayer
//
//  Created by David Chavez on 11.08.17.
//  Copyright © 2017 David Chavez. All rights reserved.
//

import Foundation
import MediaPlayer

protocol MediaWrapperDelegate: class {
    func playerUpdatedState()
    func playerSwitchedTracks(trackId: String?, time: TimeInterval?, nextTrackId: String?)
    func playerExhaustedQueue(trackId: String?, time: TimeInterval?)
    func playbackFailed(error: Error)
}

class MediaWrapper: AudioPlayerDelegate {
    private(set) var queue: [Track]
    private var currentIndex: Int
    private let player: AudioPlayer
    private var trackImageTask: URLSessionDataTask?
    
    weak var delegate: MediaWrapperDelegate?
    
    enum PlaybackState: String {
        case playing, paused, stopped, buffering, none
    }
    
    var volume: Float {
        get {
            return player.getVolume()
        }
        set {
            player.volume = newValue
        }
    }
    var rate: Float {
        get {
            return player.getRate()
        }
        set {
            player.rate = newValue
        }
    }
    
    var currentTrack: Track? {
        return queue[safe: currentIndex]
    }
    
    var bufferedPosition: Double {
        return player.currentItemLoadedRange?.latest ?? 0
    }
    
    var currentTrackDuration: Double {
        return player.currentItemDuration ?? 0
    }
    
    var currentTrackProgression: Double {
        return player.currentItemProgression ?? 0
    }
    
    var mappedState: PlaybackState {
        switch player.state {
        case .playing:
            return .playing
        case .paused:
            return .paused
        case .stopped:
            return .stopped
        case .buffering:
            return .buffering
        default:
            return .none
        }
    }
    
    
    // MARK: - Init/Deinit
    
    init() {
        self.queue = []
        self.currentIndex = -1
        self.player = AudioPlayer()
        
        self.player.delegate = self
        self.player.bufferingStrategy = .playWhenBufferNotEmpty
        
        DispatchQueue.main.async {
            UIApplication.shared.beginReceivingRemoteControlEvents()
        }
    }
    
    
    // MARK: - Public API
    
    func queueContainsTrack(trackId: String) -> Bool {
        return queue.contains(where: { $0.id == trackId })
    }
    
    func addTracks(_ tracks: [Track]) {
        queue.append(contentsOf: tracks)
    }
    
    func addTracks(_ tracks: [Track], before trackId: String?) {
        if let trackIndex = queue.index(where: { $0.id == trackId }) {
            queue.insert(contentsOf: tracks, at: trackIndex)
            if (currentIndex >= trackIndex) { currentIndex = currentIndex + tracks.count }
        } else {
            addTracks(tracks)
        }
    }
    
    func removeTracks(ids: [String]) {
        var actionAfterRemovals = "none"
        for id in ids {
            if let trackIndex = queue.index(where: { $0.id == id }) {
                if trackIndex < currentIndex { currentIndex = currentIndex - 1 }
                else if id == queue.last?.id { actionAfterRemovals = "stop" }
                else if trackIndex == currentIndex { actionAfterRemovals = "play" }
                
                queue.remove(at: trackIndex)
            }
        }
        
        switch actionAfterRemovals {
            case "play": play()
            case "stop": stop()
            default: break;
        }
    }
    
    func removeUpcomingTracks() {
        queue = queue.filter { $0.0 <= currentIndex }
    }
    
    func skipToTrack(id: String) {
        if let trackIndex = queue.index(where: { $0.id == id }) {
            currentTrack?.skipped = true
            currentIndex = trackIndex
        }
    }
    
    func playNext() -> Bool {
        if queue.indices.contains(currentIndex + 1) {
            currentIndex = currentIndex + 1
            play()
            return true
        }
        
        stop()
        return false
    }
    
    func playPrevious() -> Bool {
        if queue.indices.contains(currentIndex - 1) {
            currentIndex = currentIndex - 1
            play()
            return true
        }
        
        stop()
        return false
    }
    
    func play() {
        guard queue.count > 0 else { return }
        if (currentIndex == -1) { currentIndex = 0 }
        
        // resume playback if it was paused and check currentIndex wasn't changed by a skip/previous
        if player.state == .paused && currentTrack?.id == queue[currentIndex].id {
            player.resume()
            return
        }
        
        let track = queue[currentIndex]
        player.play(track: track)
        
        setPitchAlgorithm(for: track)
        
        // fetch artwork and cancel any previous requests
        trackImageTask?.cancel()
        if let artworkURL = track.artworkURL?.value {
            trackImageTask = URLSession.shared.dataTask(with: artworkURL, completionHandler: { (data, _, error) in
                if let data = data, let artwork = UIImage(data: data), error == nil {
                    track.artwork = MPMediaItemArtwork(image: artwork)
                }
            })
        }

        trackImageTask?.resume()
    }
    
    func setPitchAlgorithm(for track: Track) {
        if let pitchAlgorithm = track.pitchAlgorithm {
            switch pitchAlgorithm {
            case PitchAlgorithm.linear.rawValue:
                player.player?.currentItem?.audioTimePitchAlgorithm = AVAudioTimePitchAlgorithmVarispeed
            case PitchAlgorithm.music.rawValue:
                player.player?.currentItem?.audioTimePitchAlgorithm = AVAudioTimePitchAlgorithmSpectral
            case PitchAlgorithm.voice.rawValue:
                player.player?.currentItem?.audioTimePitchAlgorithm = AVAudioTimePitchAlgorithmTimeDomain
            default:
                player.player?.currentItem?.audioTimePitchAlgorithm = AVAudioTimePitchAlgorithmLowQualityZeroLatency
            }
        }
    }
    
    func pause() {
        player.pause()
    }
    
    func stop() {
        currentIndex = -1
        player.stop()
    }
    
    func seek(to time: Double) {
        self.player.seek(to: time)
    }
    
    func reset() {
        rate = 1
        queue.removeAll()
        stop()
    }
    
    // MARK: - AudioPlayerDelegate
    
    func audioPlayer(_ audioPlayer: AudioPlayer, willChangeTrackFrom from: Track?, at position: TimeInterval?, to track: Track?) {
        guard from?.id != track?.id else { return }
        delegate?.playerSwitchedTracks(trackId: from?.id, time: position, nextTrackId: track?.id)
    }
    
    func audioPlayer(_ audioPlayer: AudioPlayer, didFinishPlaying item: Track, at position: TimeInterval?) {
        if item.skipped { return }
        if (!playNext()) {
            delegate?.playerExhaustedQueue(trackId: item.id, time: position)
        }
    }
    
    func audioPlayer(_ audioPlayer: AudioPlayer, didChangeStateFrom from: AudioPlayerState, to state: AudioPlayerState) {
        switch state {
        case .failed(let error):
            delegate?.playbackFailed(error: error)
        default:
            delegate?.playerUpdatedState()
        }
    }
}
